docs/codegen.md

```diff
@@ -0,0 +1,118 @@
+
+The remaining document provides more details on generating ASTs, inspecting datatypes, tests, and information on decisions pertaining to relevant APIs.
+___
+
+### Table of Contents
+- [Generating ASTs](#Generating-ASTs)
+- [Inspecting auto-generated datatypes](#Inspecting-auto-generated-datatypes)
+- [Tests](#tests)
+- [Additional notes](#Additional-notes)
+___
+
+### Generating ASTs
+
+To parse source code and produce ASTs locally:
+
+1. Load the REPL for a given language:
+
+```
+cabal new-repl lib:tree-sitter-python
+```
+
+2. Set language extensions, `OverloadedStrings` and `TypeApplications`, and import relevant modules, `TreeSitter.Unmarshal`, `TreeSitter.Range` and `TreeSitter.Span`:
+
+```
+:seti -XOverloadedStrings
+:seti -XTypeApplications
+
+import TreeSitter.Span
+import TreeSitter.Range
+import TreeSitter.Unmarshal
+```
+
+3. You can now call `parseByteString`, passing in the desired language you wish to parse (in this case Python exemplified by `tree_sitter_python`), and the source code (in this case binary addition of two identifiers). Since the function is constrained by `(Unmarshal t, UnmarshalAnn a)`, you can use type applications to provide a top-level node `t`, an entry point into the tree, in addition to a polymorphic annotation `a` used to represent range and span:
+
+```
+parseByteString @(TreeSitter.Python.AST.Module (TreeSitter.Range.Range, Span)) tree_sitter_python "(a + b)"
+```
+
+This generates the following AST:
+
+```
+Right (Module {ann = Range {start = 0, end = 3}, extraChildren = [R1 (SimpleStatement (L1 (R1 (R1 (L1 (ExpressionStatement {ann = Range {start = 0, end = 3}, extraChildren = L1 (L1 (Expression (L1 (L1 (L1 (PrimaryExpression (L1 (L1 (L1 (R1 (BinaryOperator {ann = Range {start = 0, end = 3}, operator = L1 (R1 (R1 (L1 (Token {ann = Range {start = 1, end = 2}})))), left = PrimaryExpression (L1 (R1 (R1 (R1 (R1 (Identifier {ann = Range {start = 0, end = 1}, bytes = "a"})))))), right = PrimaryExpression (L1 (R1 (R1 (R1 (R1 (Identifier {ann = Range {start = 2, end = 3}, bytes = "b"}))))))}))))))))))) :| []}))))))]})
```

Comment:
As with other examples we’ve talked about, it might be nice to format this for the benefit of readers. (We don’t need to write code to do it or anything, just hand-formatting it would be fine.)